src/rubrix/client/datasets.py

```diff
@@ -0,0 +1,780 @@
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Dict, List, Optional, Tuple, Type, Union
+
+import pandas as pd
+from pkg_resources import parse_version
+
+from rubrix.client.models import (
+    Record,
+    Text2TextRecord,
+    TextClassificationRecord,
+    TokenAttributions,
+    TokenClassificationRecord,
+)
+from rubrix.client.sdk.datasets.models import TaskType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DatasetBase:
+    """The Dataset classes are containers for Rubrix records.
+
+    This is the base class to facilitate the implementation for each record type.
+
+    Args:
+        records: A list of Rubrix records.
+
+    Raises:
+        WrongRecordTypeError: When the record type in the provided
+            list does not correspond to the dataset type.
+    """
+
+    _RECORD_TYPE = None
+
+    def __init__(self, records: Optional[List[Record]] = None):
+        if self._RECORD_TYPE is None:
+            raise NotImplementedError(
+                "A Dataset implementation has to define a `_RECORD_TYPE`!"
+            )
+
+        self._records = records or []
+        if self._records:
+            self._validate_record_type()
+
+    def _validate_record_type(self):
+        """Validates the record type.
+
+        Raises:
+            WrongRecordTypeError: When the record type in the provided
+                list does not correspond to the dataset type.
+        """
+        record_types = {type(rec): None for rec in self._records}
+        if len(record_types) > 1:
+            raise WrongRecordTypeError(
+                f"A {type(self).__name__} must only contain {self._RECORD_TYPE.__name__}s, "
+                f"but you provided various types: {[rt.__name__ for rt in record_types.keys()]}"
+            )
+        elif next(iter(record_types)) is not self._RECORD_TYPE:
+            raise WrongRecordTypeError(
+                f"A {type(self).__name__} must only contain {self._RECORD_TYPE.__name__}s, "
+                f"but you provided {list(record_types.keys())[0].__name__}s."
+            )
+
+    def __iter__(self):
+        return self._records.__iter__()
+
+    def __getitem__(self, key):
+        return self._records[key]
+
+    def __setitem__(self, key, value):
+        if type(value) is not self._RECORD_TYPE:
+            raise WrongRecordTypeError(
+                f"You are only allowed to set a record of type {self._RECORD_TYPE} in this dataset, but you provided {type(value)}"
+            )
+        self._records[key] = value
+
+    def __delitem__(self, key):
+        del self._records[key]
+
+    def __len__(self) -> int:
+        return len(self._records)
+
+    def append(self, record: Record):
```

Comment:
Not sure if we really need this method instead of create a dataset from a list of records. What use case are you thinking about for `ds.append` instead of `records.append` + `Dataset(records)` ?

Reply:
Yeah, not sure if it will be very useful. Right now a lot of times we create an empty list and populate it in a for loop, so I thought maybe one could do the same with a dataset:
```python
dataset = rb.DatasetForText2Text():
for i in some_iterable:
    dataset.append(rb.Text2TextRecord(...))
```
Anyways, it comes more or less for free, so why not put it there ...

Reply:
Regarding the back compatibility, is easier to include a new `append` method that remove it

Reply:
I think in the tutorials we should use the workflow described above, especially if we want to have a peek at the created dataset after the loop. It's slightly more "compact" than its alternative:
```python
dataset = rb.DatasetForText2Text()
for i in some_iterable:
    dataset.append(rb.Text2TextRecord(...))
dataset.to_pandas()

# vs

records = []
for i in some_iterable:
    dataset.append(rb.Text2TextRecord(...))
dataset = rb.DatasetForText2Text(records)
dataset.to_pandas()
```

I vote for keeping the method, but if it's a show stopper for you, I'm totally fine removing it as well.

Reply:
````python
# vs
dataset = rb.DatasetForText2Text([
   rb.Text2TextRecord(...) 
   for r in some_iterable
])
dataset.to_pandas()
````

Reply:
Yeah, the examples above are a bit simplified and there is not really the need for a for loop. Here is an [example](https://rubrix.readthedocs.io/en/stable/tutorials/02-spacy.html#Using-a-Transformer-based-pipeline) where I'm not sure if you can cramp this into a list comprehension (leaving readability aside).

Anyways, i will remove it for now, but this needs a few more changes:
- [ ] update guide / docstrings
- [ ] require a list of `records` in the `Dataset*` init

Reply:
Regarding the last checkbox: We have to allow empty datasets, since `rb.load` can return empty ones. But it's a bit weird that you can create an empty `rb.DatasetFor*()`, but you cannot populate it with records.